### PR TITLE
Cel interceptor

### PR DESCRIPTION
**Pull request by bigkevmcd**
# Changes

This implements a [CEL](https://github.com/google/cel-go) based interceptor that allows matching on request bodies, and request headers.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [X] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [X] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/triggers/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

```
This adds a new CEL interceptor, able to use expressions to determine whether or not to filter out incoming request bodies.
```

### CODE REVIEW

**Comment by bigkevmcd** [comment]
Some thoughts I've had since the initial implementation...

Instead of one `expression` there could be several.

```yaml
expressions:
 - "headers.match(...)"
 - "body.pull_request.head.repo.fork == false"
```

Where each of the expressions would need to return true, this would improve readability (over very long expressions).

Also, there's no way currently to add a custom response header, this could be implemented similar to the `values:` key.

**Comment by bigkevmcd** [comment]
Also, in order to avoid the invisible coupling between the interceptor modifying the request body, and bindings, CEL could be added to the list of expression languages for bindings.

File: pkg/apis/triggers/v1alpha1/event_listener_types.go [hunk 1]
```diff
@@ -105,6 +106,12 @@ type GitlabInterceptor struct {
 	EventTypes []string   `json:"eventTypes,omitempty"`
 }
 
+// CELInterceptor provides a webhook to intercept and pre-process events
+type CELInterceptor struct {
+	Expression string            `json:"expression,omitempty"`
```
Comment (dibyom) [inline pkg/apis/triggers/v1alpha1/event_listener_types.go]: maybe we can call this `filterExpression` or `filters`? More explicit than the `Value` field which are also technically CEL expressions!

File: pkg/apis/triggers/v1alpha1/event_listener_validation.go [hunk 2]
```diff
@@ -67,7 +67,7 @@ func (s *EventListenerSpec) validate(ctx context.Context, el *EventListener) *ap
 
 func (i *EventInterceptor) validate(ctx context.Context, namespace string) *apis.FieldError {
 	// Validate at least one
-	if i.Webhook == nil && i.Github == nil && i.Gitlab == nil {
+	if i.Webhook == nil && i.Github == nil && i.Gitlab == nil && i.CEL == nil {
```
Comment (dibyom) [inline pkg/apis/triggers/v1alpha1/event_listener_validation.go]: Should we also validate that if CEL has the `expression` field set?

File: pkg/interceptors/cel/cel.go [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+		KubeClientSet:          k,
+		EventListenerNamespace: ns,
+	}
+}
+
+// ExecuteTrigger is an implementation of the Interceptor interface.
+func (w *Interceptor) ExecuteTrigger(payload []byte, request *http.Request, _ *triggersv1.EventListenerTrigger, _ string) ([]byte, error) {
+	env, err := makeCelEnv()
+	if err != nil {
+		return nil, err
```
Comment (dibyom) [inline pkg/interceptors/cel/cel.go]: Could we add some extra context when returning errors in this file? 
e.g. `fmt.Errof("error creating cel environment: %w, err)`

File: pkg/interceptors/cel/cel_test.go [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+			args: args{
+				payload: []byte(`{}`),
+			},
+			want:    []byte(`{}`),
+			wantErr: false,
+		},
+		{
+			name: "simple header check with non matching header",
+			CEL: &triggersv1.CELInterceptor{
+				Expression: "headers['X-Test'][0] == 'unknown'",
```
Comment (dibyom) [inline pkg/interceptors/cel/cel_test.go]: The `headers` part here is slightly different from `headers` in `TriggerBindings`. 
In Bindings, the keyword is `header` not `headers` -- though `headers` does sound like the more correct term.
Also, you do not have to use array accessors e.g.  `$(headers.foo)` not `$(headers.foo[0])`
Its probably worth being consistent here.

**Comment by dibyom** [comment]
> Instead of one expression there could be several.

Another more verbose option could be the multiple interceptor support that @wlynch is working on

> Also, in order to avoid the invisible coupling between the interceptor modifying the request body, and bindings, CEL could be added to the list of expression languages for bindings

This could be an interesting add on for the future...(JSONPath also supports filtering, maybe that could be a interceptor too :stuck_out_tongue: )

**Comment by bigkevmcd** [comment]
> > Instead of one expression there could be several.
> 
> Another more verbose option could be the multiple interceptor support that @wlynch is working on

This would be slightly different, in that it would be more about readability to avoid having to write long expressions like `filter1 && filter2 && filter3 && filter4`

File: pkg/interceptors/cel/cel_test.go (hunk 4 above)
Comment (bigkevmcd) [reply pkg/interceptors/cel/cel_test.go]: Just to clarify, you would prefer it to emulate JSON path expressions in CEL?

Comment (dibyom) [reply pkg/interceptors/cel/cel_test.go]: Yes!

File: pkg/interceptors/cel/cel.go (hunk 3 above)
Comment (bigkevmcd) [reply pkg/interceptors/cel/cel.go]: Done

File: pkg/apis/triggers/v1alpha1/event_listener_validation.go (hunk 2 above)
Comment (bigkevmcd) [reply pkg/apis/triggers/v1alpha1/event_listener_validation.go]: Add validation for the renamed `filter` expression.

**Comment by bigkevmcd** [comment]
The main driver for this is to get rid of an external interceptor that is doing matching, and extracting the SHA and trimming it (amongst other things).

The "values" behaviour emulates this behaviour, so, I'd like to see some discussion about what alternatives could look like.

File: pkg/interceptors/cel/cel_test.go (hunk 4 above)
Comment (bigkevmcd) [reply pkg/interceptors/cel/cel_test.go]: I'm wondering if CEL is the right expression language to use, maybe we could augment the JSON path expressions?

Comment (bigkevmcd) [reply pkg/interceptors/cel/cel_test.go]: To be clear, the syntax `header['X-Test'][0]` is one possible option.

The header value is a https://golang.org/pkg/net/http/#Header value, but, it's translated in CEL into a map, so, there's a custom match overload, which uses `.Get()` on the value, this is doing more than a simple key-value lookup.

I'm not sure how easy it is to emulate JSON path expressions in CEL without a significant amount of work.

File: pkg/interceptors/cel/cel_test.go [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+		{
+			name: "simple body check with matching body",
+			CEL: &triggersv1.CELInterceptor{
+				Filter: "body.value == 'testing'",
+			},
+			args: args{
+				payload: []byte(`{"value":"testing"}`),
+			},
+			want:    []byte(`{"value":"testing"}`),
+			wantErr: false,
```
Comment (wlynch) [inline pkg/interceptors/cel/cel_test.go]: Mixing both error and non-error tests makes it a little harder to follow which cases should and should not work. Can we split these cases into 2 separate tests?

File: pkg/interceptors/cel/cel_test.go [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+
+// Allow configuration via a config map
+
+func TestInterceptor_ExecuteTrigger(t *testing.T) {
+	type args struct {
+		payload []byte
+	}
+	tests := []struct {
+		name    string
+		CEL     *triggersv1.CELInterceptor
```
Comment (wlynch) [inline pkg/interceptors/cel/cel_test.go]: It looks like the only thing that changes here is the filter string. The test case struct would be simpler if we just use `filter string` here and create the CELInterceptor in the test.

File: pkg/interceptors/cel/cel_test.go [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+		args    args
```
Comment (wlynch) [inline pkg/interceptors/cel/cel_test.go]: Similarly, `args` is just a wrapper type for `[]byte`. Can we use the type directly here?

File: pkg/sink/sink.go [hunk 8]
```diff
@@ -92,6 +93,8 @@ func (r Sink) HandleEvent(response http.ResponseWriter, request *http.Request) {
 				interceptor = github.NewInterceptor(t.Interceptor.Github, r.KubeClientSet, r.EventListenerNamespace, log)
 			case t.Interceptor.Gitlab != nil:
 				interceptor = gitlab.NewInterceptor(t.Interceptor.Gitlab, r.KubeClientSet, r.EventListenerNamespace, log)
+			case t.Interceptor.CEL != nil:
```
Comment (wlynch) [inline pkg/sink/sink.go]: Can we add a test at the interceptor level to make sure things work as expected end to end?

File: pkg/interceptors/cel/cel_test.go (hunk 4 above)
Comment (dibyom) [reply pkg/interceptors/cel/cel_test.go]: By "emulating", I was just referring to making it easy to access single valued headers i.e header['X-Test'][0] vs header['X-Test']. From reading @wlynch 's comment it seems like that might already be the case...
We definitely should not try to do anything more complicated than that 😅

File: pkg/interceptors/cel/cel.go [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+
+	out, _, err := prg.Eval(data)
+	return out, err
+}
+
+func embeddedFunctions() cel.ProgramOption {
+	return cel.Functions(
+		&functions.Overload{
+			Operator: "match",
+			Function: matchHeader},
```
[the remark this answers is not in the document]
Comment (bigkevmcd) [reply pkg/interceptors/cel/cel.go]: That's not quite what it does, the value of `header` is https://golang.org/pkg/net/http/#Header and the `match` delegates to the `Get` method, which does a bit more than accessing the header for single-value headers.

File: pkg/interceptors/cel/cel.go [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+
+	}
+
+	return payload, nil
+}
+
+func evaluate(expr string, env cel.Env, data map[string]interface{}) (ref.Val, error) {
+	parsed, issues := env.Parse(expr)
+	if issues != nil && issues.Err() != nil {
+		return nil, issues.Err()
```
[the remark this answers is not in the document]
Comment (bigkevmcd) [reply pkg/interceptors/cel/cel.go]: The error message from CEL is pretty good, the test I added only really tests a subset of the message.

```
failed to evaluate expression 'header['X-Test': ERROR: <input>:1:8: Syntax error: token recognition error at: ''X-Test'
 | header['X-Test
 | .......^
ERROR: <input>:1:15: Syntax error: mismatched input '<EOF>' expecting {'[', '{', '(', '.', '-', '!', 'true', 'false', 'null', NUM_FLOAT, NUM_INT, NUM_UINT, STRING, BYTES, IDENTIFIER}
 | header['X-Test
 | ..............^
```

File: pkg/apis/triggers/v1alpha1/event_listener_validation.go [hunk 11]
```diff
@@ -67,7 +67,7 @@ func (s *EventListenerSpec) validate(ctx context.Context, el *EventListener) *ap
...
 		return apis.ErrMissingField("interceptor")
 	}
```
[the remark this answers is not in the document]
Comment (dibyom) [reply pkg/apis/triggers/v1alpha1/event_listener_validation.go]: I think this still needs to be added

**Comment by dibyom** [comment]
> We should make sure to add a reference to the issue (#49) in the PR body / commit message.

Looks like we need to add this as well!

File: docs/eventlisteners.md [hunk 12]
```diff
@@ -194,3 +194,32 @@ spec:
       template:
         name: pipeline-template
 ```
+
+### CEL Interceptors
+
+CEL interceptors parse expressions to filter requests based on JSON bodies and request headers, using the
+[CEL](https://github.com/google/cel-go) expression language.
+
+Supported features include case-insensitive matching on request headers.
```
Comment (wlynch) [inline docs/eventlisteners.md]: Let's document the custom methods we add similar to https://github.com/google/cel-spec/blob/master/doc/langdef.md#list-of-standard-definitions

Comment (bigkevmcd) [reply docs/eventlisteners.md]: Is this ok for a start?

https://github.com/tektoncd/triggers/blob/5d0379587187bef6ced05c6469b10d272d28009f/docs/cel_expressions.md

Comment (wlynch) [reply docs/eventlisteners.md]: Looks good! :smile: Thanks!

**Pull request merged**